File: May-LeetCoding-Challenge/19-Online-Stock-Span/Online-Stock-Span.cpp
/*
    Dynamic programming using idea of KMP.
    Got AC using array.
    Got TLE using vector.
*/ 
class StockSpanner {
public:
    int fail[10005];
    int v[10005];
    int n;
    StockSpanner() {
        n = 1;
        fail[0] = 0;
    }
    
    int next(int price) {
        v[n++] = price; 
        int i = n - 2;
        while (i > 0 and price >= v[i]) i = fail[i];
        fail[n-1] = i;
        return n - i - 1;
          
    }
};


/*
Stack !
*/
class StockSpanner2 {
public:
    stack<pair<int, int>> st; 
    StockSpanner() {}
    
    int next(int price) {
        int sum = 1;
        
        while (!st.empty() and price >= st.top().first) {
            sum += st.top().second;
            st.pop();
        }
        st.push(make_pair(price, sum));
        return sum;
    }
};

File: May-LeetCoding-Challenge/19-Online-Stock-Span/Online-Stock-Span.kt
import java.util.*

/**
 * Your StockSpanner object will be instantiated and called as such:
 * var obj = StockSpanner()
 * var param_1 = obj.next(price)
 */
class OnlineStockSpanKotlin901 {
    // 100 80 60 70 60 75 85
    // 1    1 1  2  1  4  6

    // 100 1    add 100
    // 100 1 -> 80 1    add 80
    // 100 1 -> 80 1 -> 60 1    add 60
    // 100 1 -> 80 1 -> 70 2    add 70
    // 100 1 -> 80 1 -> 70 2 -> 60 1    add 60
    // 100 1 -> 80 1 -> 75 4    add 75
    // 100 1 -> 85 6    add 85

    // 1384 ms
    // first price, second count
    private val priceCountStack: Stack<Pair<Int, Int>> = Stack()
    fun next(price: Int): Int {
        var count = 1
        // stack peek won't remove the element, just return the reference of last added element
        while (priceCountStack.isNotEmpty() && priceCountStack.peek().first <= price) {
            count += priceCountStack.pop().second
        }
        priceCountStack.push(Pair(price, count))
        return count
    }

    /*
    1884 ms
    private val prices: MutableList<Int> = ArrayList()
    fun next(price: Int): Int {
        var count = 1
        loop@ for (index in prices.size - 1 downTo 0) {
            when {
                prices[index] <= price -> ++count
                else -> break@loop
            }
        }
        prices.add(price)
        return count
    }
     */
}

fun main() {
    val solution = OnlineStockSpanKotlin901()
    // 1112146
    print(solution.next(100))
    print(solution.next(80))
    print(solution.next(60))
    print(solution.next(70))
    print(solution.next(60))
    print(solution.next(75))
    print(solution.next(85))
}